rfd/0078-login-rules.md

```diff
@@ -0,0 +1,504 @@
+  priority: 0
+
+  # traits is a single predicate expression which must return a dict which will
+  # set the user's traits during login
+  traits: >
+    dict(
+      pair("groups",
+        ifelse(external.groups.contains("splunk"),
+          external.groups.add("dbs"),
+          external.groups)),
+      pair("logins",
+        set(
+          "ubuntu",
+          ifelse(external.username.contains("nic@goteleport.com"), "root", ""),
+          ifelse(external.organization.contains("teleport"), "teleporters", "external"))))
+
+  # eventually login_rules could also return the desired teleport roles for the
+  # user to allow custom logic, but this will be out of scope for the initial
+  # implementation
+  # roles: >
+  #   set("ssh-users", "db-users")
+```
+
+### Predicate Helper Functions
+
+A set of helper functions to be used in the predicate expressions will be
+included.
+Part of the reasoning for using the
+[predicate language](https://github.com/vulcand/predicate)
+is so that it will be extensible.
+As the need arises, we can always add new helper function.
+
+The predicate parser is based on the `Go` language parser, so `Go` keywords
+which may be better names (such as `if` or `select`) must unfortunately be avoided.
+
+#### `set(...items)`
+
+`set()` returns a set including all its arguments.
+Duplicates are not possible and will be filtered out on creation of the set if
+they are passed in.
+All items must be strings.
+As a special case, the empty string will not be included in the set.
+
+#### `set.contains(value)`
+
+`set.contains(value)` returns `true` if any item in the set is an exact match
+for `value`, else it returns `false`.
+`set("a", "b").contains("b")` returns `true`.
+
+#### `set.add(value)`
+
+`set.add(value)` returns a copy of the set with `value` added.
+`set("a", "b").add("c")` returns `("a", "b", "c")`.
+
+#### `set.remove(value)`
+
+`set.remove(value)` returns a copy of the set with `value` removed.
+`set("a", "b", "c").remove("c")` returns `("a", "b")`.
+
+#### `dict(...pairs)`
+
+`dict()` returns a dictionary with keys of type `string` and values of type `set`.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+)
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("asparagus", "brocolli"),
+}
+```
+
+Arguments must be pairs, the first element will be the key,
+and the second element will be the value.
+
+#### `dict.add_values(key, ...values)`
+
+`dict.add_values(key, ...values)` returns a copy of the dict with the given values
+added to the set at `dict[key]`.
+If `dict[key]` is empty or it does not exist, it will be added with `values` as
+its only elements.
+
+```
+dict(
+  pair("fruits", set("apple")),
+).add_values("fruits", "banana").add("vegetables", "asparagus", "brocolli")
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("asparagus", "brocolli"),
+}
+```
+
+#### `dict.remove_keys(...keys)`
```

Comment:
switch to just `remove` but imo it might as well support multiple keys. A known customer usecase will be removing many traits, `remove("a", "b", "c")` seems nicer than `.remove("a").remove("b").remove("c")`

Reply:
Sounds good.